xc/device.ml

```diff
@@ -1743,43 +1743,70 @@ module Vusb = struct
     else
       []
 
-  let vusb_controller_plug ~xs ~domid ~driver ~driver_id =
-    if (Qemu.is_running ~xs domid)
-    && not (List.mem driver_id (qom_list ~xs ~domid))then
-      qmp_send_cmd domid Qmp.(Device_add Device.({driver; device=USB { USB.id=driver_id; params=None }})) |> ignore
 
-  let vusb_plug ~xs ~privileged ~domid ~id ~hostbus ~hostport ~version =
+  let vusb_plug ~xs ~privileged ~domid ~id ~hostbus ~hostport ~version ~speed =
     debug "vusb_plug: plug VUSB device %s" id;
-    let get_bus v =
-      if String.startswith "1" v then
-        "usb-bus.0"
-      else begin
-        (* Here plug usb controller according to the usb version*)
-        let usb_controller_driver = "usb-ehci" in
-        let driver_id = "ehci" in
-        vusb_controller_plug ~xs ~domid ~driver:usb_controller_driver ~driver_id;
-        Printf.sprintf "%s.0" driver_id;
-      end
+    let get_bus () =
+      let vusb_controller_plug ~xs ~domid ~driver ~driver_id =
+        (* requires Qemu.is_running *)
+        if not (List.mem driver_id (qom_list ~xs ~domid)) then
+          qmp_send_cmd domid Qmp.(Device_add Device.({driver; device=USB { USB.id=driver_id; params=None }})) |> ignore
+      in
+      let usb_bus0 = "usb-bus.0", fun () -> () in
+      let ehci0 = "ehci.0", fun () -> vusb_controller_plug ~xs ~domid ~driver:"usb-ehci" ~driver_id:"ehci" in
+      let speed_of_float x = if x <= 0. then
+                               `Unknown
+                             else if x <= 1.5 then
+                               `Low    (* v1.0 *)
+                             else if x <= 12. then
+                               `Full   (* v1.1 *)
+                             else if x <= 480. then
+                               `High   (* v2.0 *)
+                             else if x <= 5000. then
+                               `Super  (* v3.0 *)
+                             else
+                               `Unknown
+      in
+      let bus_from_speed =
+        match float_of_string_opt speed with
```

Comment:
See https://github.com/xapi-project/xen-api/pull/3984#discussion_r358164501 this conversion can be avoided.